pkg/util/util_supported.go

```diff
@@ -89,3 +91,28 @@ func GetRuntimeDir() (string, error) {
 	}
 	return rootlessRuntimeDir, nil
 }
+
+// ReadPassphraseInCLIParsingPhase reads a password from the terminal. On interruption or termination signal, it will restore the terminal and terminate the program
+func ReadPassphraseInCLIParsingPhase(fd int) ([]byte, error) {
+	// Store and restore the terminal status on interruptions to
+	// avoid that the terminal remains in the password state
+	// This is necessary as for https://github.com/golang/go/issues/31180
+
+	oldState, _ := terminal.GetState(fd)
+	interruptChannel := make(chan os.Signal, 1)
+	signal.Notify(interruptChannel, syscall.SIGINT, syscall.SIGTERM)
+	defer func() {
+		signal.Stop(interruptChannel)
+		close(interruptChannel)
+	}()
+	go func(oldState *terminal.State) {
+		for range interruptChannel {
+			if oldState != nil {
+				terminal.Restore(fd, oldState)
+			}
+			signal.Reset(os.Interrupt)
+			os.Exit(1)
```

Comment:
Why not just return an error and give Podman a chance to properly exit?  There is usually some cleanup do be done before Podman exists.

Reply:
Because there is no way to return from `terminal.ReadPassword`, so this would mean the terminal remains blocked.

Reply:
So if you returned the error, users would need to hit <return> or if you exit here users have to hit <return> correct?

Reply:
Not sure that I understand you - If we return a error, the stdin would still be blocked by `terminal.ReadPassword` until users hit return.

We even [tried to](https://github.com/containers/common/pull/1312#discussion_r1099038041) close the fd but not even that unblocks the `read` call from `terminal.ReadPassword` . So returning an error here on interrupt is possible, but it would mean that the terminal remains in a weird state, which is likely not usable on subsequent calls. :slightly_frowning_face:

Reply:
Have you experimented with returning, and seen what happens from the user point of view?

Reply:
Yes and this is interesting because it revealed a race condition. There is somewhere else a signal handler for the interrupt, which terminates `podman` with "exit status 1". We would need to disable this signal handler for this function to work properly, which is ugly, but it would work. From the user's perspective there is no noticeable change, except that the return code would be 125 instead of 1.

Should we switch to that approach? It's ugly because we would need to reset the SIGINT signal handler and in the error case the terminal is blocked until the program terminates, but the alternative is to use `os.Exit(1)` within the function, which is nasty for other reasons.

I have no preference and am happy to continue with either way :slightly_smiling_face:

Reply:
The SIGINT handler in question should be https://github.com/containers/podman/blob/main/libpod/shutdown/handler.go but that doesn't help us much - that's nestled deep into Libpod and as such we can't touch it from here in c/common.

Generally speaking we should avoid `os.Exit()` whenever possible (for the exact reasons we wrote the shutdown handler in Libpod - there are situations where we need to keep Podman alive long enough to finish essential cleanup before obeying user-initiated shutdown), so I'm in favor of returning an error, but I have no easy way of disabling the interrupt handler from within c/common to give us consistent error codes.

Reply:
I see that almost everyone in this threads favours an approach of returning an error. This doesn't remove the race condition with [libpod/shutdown/handler.go](https://github.com/containers/podman/blob/0b612a9a2d812308e892dd909c53b4b2e0cfde3d/libpod/shutdown/handler.go#L45C57-L45C57), but it works in most of the cases I tested, so it would improve the situation already substantially.

I will update the PR accordingly.

Reply:
So I think the answer here is to take a step back, I understand what this PR is trying to do and that looks reasonable. 

If we look at the code path exactly `podman login` bypasses pretty much all of libpod and calls directly in to pkg/auth here. So shouldn't be the answer disable the libpod setup for podman login/logout in which case there cannot be a race between the signal handlers.